posix: options: add XSI_REALTIME option group

cfriedt opened the pull request:
Add a POSIX Option Group called `XSI_REALTIME` (with Kconfig option `CONFIG_XSI_REALTIME`).
    
When `XSI_REALTIME` is selected (or when required POSIX Options are enabled), define `_XOPEN_REALTIME` to be something other than -1 (`_XOPEN_VERSION` seemed appropriate).

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/83303/docs/services/portability/posix/index.html)

Closes #82004
Relates to #79816

Required changes:

- [x] #83651
- [x] #83645
- [x] #83646
- [x] #83647
- [x] #83648
- [x] #83649
- [x] #83650
- [x] #83708
- [x] #84349
- [x] #84354
- [x] #84668
- [x] #85834
- [x] #85837

Commit:
posix: options: split kconfig.xsi into separate option groups

Since XSI is composed of several distinct POSIX Option Groups
split Kconfig.xsi into separate files - one for each Option
Group.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: options: add xsi realtime option group

Add a POSIX Option Group called XSI_REALTIME (with Kconfig
option CONFIG_XSI_REALTIME).

When XSI_REALTIME is selected (or when required POSIX Options
are enabled), define _XOPEN_REALTIME to be something other
than -1 (_XOPEN_VERSION seemed appropriate).

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: options: profiles: uncomment options that should be selected

For the REALTIME_MINIMAL Application Environment Profile, uncomment
the POSIX_MEMLOCK, POSIX_MEMLOCK_RANGE, POSIX_MEMORY_PROTECTION,
POSIX_MAPPED_FILES, and POSIX_SHARED_MEMORY_OBJECTS options.

These should have been uncommented back when Kconfig options for
those features were added, so this can probably be called a
Kconfig bug.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

ycsin commented on lib/posix/options/mlock.c:
```diff
@@ -9,16 +9,22 @@
 int mlock(const void *addr, size_t len)
 {
-	k_mem_pin(addr, len);
+	void *const _addr = (void *)addr;
```
the (2nd) commit msg is inaccurate

cfriedt replied:
Fixed!

ycsin commented on lib/posix/options/mlockall.c:
```diff
@@ -9,6 +9,7 @@
 #include <sys/types.h>
 
 #include <zephyr/posix/sys/mman.h>
+#include <zephyr/toolchain.h>
```
the (4th) commit msg is inaccurate

cfriedt replied:
Fixed!

ycsin commented on lib/posix/options/Kconfig.mem:
```diff
@@ -50,7 +50,7 @@ config POSIX_MEMLOCK
 	bool "POSIX range memory locking"
 	imply MMU if (CPU_HAS_MMU && ARCH_HAS_DEMAND_PAGING)
-	imply DEMAND_PAGING
+	imply DEMAND_PAGING if !(X86 && !X86_64 && USERSPACE)
```
should we create an enhancement issue for these exclusions (when this PR is merged)?

cfriedt replied:
Probably some kind of issue

ycsin reviewed (dismissed):
LGTM!

cfriedt commented:
Reviewers might want to take a look workarounds for some weird corner cases

* demand paging / default ram backend instead of no implementation (@npitre)
* xtensa linker script syntax error (@ceolin)
* 32-bit x86 and userspace issue (@dcpleung)

The latter two should be reproducible by undoing the workaround commits.

cfriedt commented:
I'll see if I can separate this PR into smaller, more manageable ones.

Currently, it's blocking #81859 and should have reviews by @dcpleung @ceolin and @npitre on the topics mentioned above.

dcpleung commented:
I guess the `z_x86_kpti_is_access_ok()` extern declaration is missing a semi-colon at the end?

cfriedt commented:
> I guess the `z_x86_kpti_is_access_ok()` extern declaration is missing a semi-colon at the end?

@dcpleung - It was surprisingly hard to see that on mobile for some reason 😅

Commit:
doc: posix: mark XSI_REALTIME as supported

Mark the XSI_REALTIME Option Group as supported, such that the
_XOPEN_REALTIME feature test macro may be tested.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented on lib/posix/options/Kconfig.xsi_streams:
```diff
@@ -0,0 +1,12 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config XOPEN_STREAMS
```
This will need to be deprecated and renamed `XSI_STREAMS`

cfriedt commented on lib/posix/options/Kconfig.xsi_single_process:
```diff
@@ -0,0 +1,14 @@
+
+config XSI_SINGLE_PROCESS
+	bool "X/Open single process"
+	depends on POSIX_SINGLE_PROCESS
+	depends on POSIX_TIMERS
```
I think that XSI_SINGLE_PROCESS should select these other options, similarly to how it's done for XSI_REALTIME. That's a change for another PR though.

cfriedt commented:
@ycsin - should be green shortly (finally 😅)

nordicjm commented on lib/posix/options/CMakeLists.txt:
```diff
@@ -75,10 +75,6 @@ if (NOT CONFIG_TC_PROVIDES_POSIX_FILE_SYSTEM_R)
-zephyr_library_sources_ifdef(CONFIG_POSIX_MEMLOCK mlockall.c)
-zephyr_library_sources_ifdef(CONFIG_POSIX_MEMLOCK_RANGE mlock.c)
-
 if (NOT CONFIG_TC_PROVIDES_POSIX_MEMORY_PROTECTION)
 zephyr_library_sources_ifdef(CONFIG_POSIX_MEMORY_PROTECTION mprotect.c)
```
```suggestion
if(NOT CONFIG_TC_PROVIDES_POSIX_MEMORY_PROTECTION)
  zephyr_library_sources_ifdef(CONFIG_POSIX_MEMORY_PROTECTION mprotect.c)
```
(for a future PR)

cfriedt replied:
Good catch - we should probably have a cmake linter in GH actions

cfriedt replied:
@nordicjm - created https://github.com/zephyrproject-rtos/zephyr/issues/84876

nordicjm commented on lib/posix/options/CMakeLists.txt:
```diff
@@ -154,6 +142,15 @@ if (NOT CONFIG_TC_PROVIDES_POSIX_THREADS)
   )
 endif()
 
+if (NOT CONFIG_TC_PROVIDES_XSI_REALTIME)
```
```suggestion
if(NOT CONFIG_TC_PROVIDES_XSI_REALTIME)
```

cfriedt replied:
Fixed!

ycsin reviewed (approved):
LGTM

cfriedt commented:
@nordicjm - would you care to give a second approval?
